File: python/py_algorithms/first_fit_decreasing.py
def first_fit_decreasing(available, required):
    sorted_required = sorted(required, reverse=True)
    bins = []
    waste = 0
    stock_used = 0

    for item in sorted_required:
        placed = False
        for bin in bins:
            if bin['remaining'] >= item:
                bin['cuts'].append(item)
                bin['remaining'] -= item
                placed = True
                break
        if not placed:
            new_bin = {
                'cuts': [item],
                'remaining': available - item
            }
            bins.append(new_bin)
            stock_used += 1

    waste = sum(bin['remaining'] for bin in bins)
    return {'cut_list': bins, 'waste': waste, 'stock_used': stock_used}

File: JS/JS_algorithms/firstFitDecreasing.js
function firstFitDecreasing(available, required) {
    let sortedRequired = [...required].sort((a, b) => b - a);
    let bins = [];
    let stockUsed = 0;

    sortedRequired.forEach(item => {
        let placed = false;

        for (let bin of bins) {
            if (bin.remaining >= item) {
                bin.cuts.push(item);
                bin.remaining -= item;
                placed = true;
                break;
            }
        }

        if (!placed) {
            let newBin = { cuts: [item], remaining: available - item };
            bins.push(newBin);
            stockUsed++;
        }
    });

    let waste = bins.reduce((acc, bin) => acc + bin.remaining, 0);
    return { cutList: bins, waste, stockUsed };
}

module.exports = firstFitDecreasing;
